### ISSUE TITLE
Janus mistakenly ignores trickle candidates if the candidates are received when cleaning up a previous ice session

### ISSUE THREAD

**Issue by tugtugtug** [meetecho/janus-gateway#1026]
Steps to reproduce:
1. have a client publishing video/audio to the videoroom.
2. unpublish
3. publish again in a short period of time.

What happens here is in janus.c we have the sdp being held up due to JANUS_ICE_HANDLE_WEBRTC_CLEANING being set, however the subsequent tickle ice candidates received are being send to the ice session that is cleaning up and effectively ignored. Since there is no error reported either way, we end up with a broken ice state.

I think we should clean up the pending candidates list as soon as we receive a new offer and allow caching the candidates from then on.

> If you want to open an issue, please make sure that:
> 
> 1. you actually found a bug: for generic questions, use the [meetecho-janus](http://groups.google.com/d/forum/meetecho-janus) group instead;
> 2. nobody opened the same bug already (do a search);
> 3. the issue wasn't already solved;
> 4. you don't paste a huge amount of text in the issue or comments: use a service like [pastebin](http://pastebin.com/) or similar;
> 5. you provide a GDB stacktrace if Janus crashed on you, and/or output from AddressSanitizer or Valgrind (more details available[here](http://janus.conf.meetecho.com/docs/debug.html)).
> 
> If so, just remove this text and share your findings!

**Comment by lminiero** [comment]
I think this is part of the fixes @atoppi is working on in this branch here: https://github.com/meetecho/janus-gateway/tree/iceloop-sessions-mgmt (not in a PR yet)
If that's not working yet, please sync with him about possible solutions to this issue.

**Comment by atoppi** [comment]
I'm wondering how to deal with this kind of scenario from a Janus API perspective.
I mean, the CLEANING flag could be possibly triggered in such cases:
1) DTLS handshake error
2) `janus_ice_webrtc_hangup`
3) `janus_ice_thread` leaving
The flag is cleared once `janus_ice_webrtc_free` is called, and this function is the one in charge of deleting pending trickles.
So I think that Janus should reject any trickles incoming while doing its internal cleanups and these cleanups should be as fast as possibile (that's why I'm removing the sleep calls).

**Comment by tugtugtug** [comment]
I think it is a very common and useful case to allow quickly  unpublish and republish. We have this case to switch between desktop sharing and camera for example and is hitting this bug every single time. With your change to reject the newer jsep will simply make the client keep trying, which is not quite an ideal scenario for either the client or the server.
While sleep is indeed not ideal, we could however use states to solve this, in my quick change I've added a new state and if we don't want to sleep, then we could potentially cache both the jsep and the trickle candidates until the ice session is cleaned up.
Another thought is maybe we could think about allowing a new ice session while the stake one is cleaning up?

**Comment by lminiero** [comment]
Quickly unpublishing/republishing is not currently supported anyway. While Alessandro's patch will help, you cannt currently republish if a second hasn't passed yet. Check the reference counters branch (#403) for something that doesn't suffer from the same issue: ideally we'll merge that soon, so the more feedback we get the better.

**Comment by atoppi** [comment]
In your specific case, the `unpublish` is directly calling `janus_ice_webrtc_hangup`. This sets the flag CLEANING and calls `g_main_loop_quit`, triggering the teardown of `janus_ice_thread`. During its last procedures the thread invokes `janus_ice_webrtc_free`, finally and clears out the CLEANING flag.

In https://github.com/meetecho/janus-gateway/tree/iceloop-sessions-mgmt branch the sleeps have been removed on the described chain of calls, but still persists one asynchronous step (the `g_main_loop_quit` essentially schedules the icethread leaving for the next tick of operation). That is for the nature of glib main loops. This means that while the client will get the answer for the `unpublish`, the real cleaning ends up as soon as the icethread is able to terminate. Ideally this time difference is very tiny, and should not be a problem even for quick unpublish/publish jobs.
Give a try to https://github.com/meetecho/janus-gateway/tree/iceloop-sessions-mgmt to see if that can help in your scenario.

**Comment by tugtugtug** [comment]
Thanks Folks, a second thought,
 It seems to have unpublish being an asynchronous API, and return an ack first, and return success only after the ICE session is cleaned up is the right solution, and would fit perfectly with what Alessandro has done, and allow client to react properly (i.e. wait).
 To avoid backward compatibility, we might need to have an "unpublish_async" or "unpublish_callback" added.
 I guess a request down the road is to support re-inivite, so we don't always have to unpublish/re-publish.

**Comment by atoppi** [comment]
@tugtugtug looking at your alternative fix I've come up with a [more conservative approach](https://github.com/meetecho/janus-gateway/commit/e2b19e26d4c042e0c1c64ac10e99472f3b776ad7).
I want to delay the `janus_ice_notify_hangup` (a.k.a. notify the client) until `janus_ice_webrtc_free` is invoked, and this is obtained by exploiting a `hangup_reason` string into the handle, just like you did.
On the other side the `unpublished` event still has the same semantic and timing as before: it is generated before an actual hangup occurs so it does not grant that the WebRTC stack has been cleaned but it does grant that the publisher has been removed from the room.
In this way you can use the `unpublished` event to handle your UI and notify the event, while the subsequent `hangup` event will grant you that the server is ready for another `publish` request.

I think this approach should deal with your request of "changing unpublish to respond after the ice session is cleaned up", by not adding so much complexity and not involving the storage of pending trickles/jseps.

**Comment by tugtugtug** [comment]
@atoppi Thanks for taking this into account. I think it is workable with your change to add logic to listen to the hangup event. I still feel it a bit awkward, as the unpublish response really is not useful as it is now, it pretty much as the same value as the ack.

**Comment by atoppi** [comment]
@tugtugtug well try to see it from this perspective: even the setup API (`publish`/`configure`) receives a response (`configured`) that is asynchronous with the WebRTC connection state. Now we have just the same behavior for setup and teardown.

**Comment by tugtugtug** [comment]
@atoppi true, but not quite the same, the setup API as you mentioned returns early as well, but the code properly handles trickle candidates before the actual setup is completed, which works as if the API returned after the setup work is done. The fact that currently the unpublish response returns and re-publish immediately would always fail is what makes it awkward and inconsistent with the setup API.

**Comment by lminiero** [comment]
Immediate re-publish is not meant to work in master. Trying to do that will always fail. Refer to the refcount branch if you need consistency.

**Comment by tugtugtug** [comment]
Thanks @lminiero, looks like that branch has existed for a while, any plan to merge it back to master or you prefer it there until more soaking is done?

**Comment by lminiero** [comment]
The only reason it hasn't been merged yet is that we're ironing out some last issues. But until more people will start playing with is that moment will be delayed. Considering you're trying to solve an issue that branch was meant to fix since the beginning, I think it makes more sense to focus on that, rather than hammering in the current master stuff that we'll get rid of anyway later on.

**Comment by tugtugtug** [comment]
Sounds good, we'll start running some testing on the branch and see how it goes.

**Comment by lminiero** [comment]
Any updates on the testing?

**Comment by tugtugtug** [comment]
@lminiero we had to switch our focus the last few weeks, so we haven't got much chance to verify this yet. Will try to get an update before the holidays.

**Comment by tugtugtug** [comment]
@lminiero Not much luck, create an audiobridge and join failed. Seems the refcount for the room reaches 0  right after it is created. Logs at https://pastebin.com/pD12RD5y

**Comment by tugtugtug** [comment]
also captured a backtrace here. https://pastebin.com/iNkAhBsE
Looks like the insert gets called twice, and the 2nd one updates the value, thus the destroy.

**Comment by tugtugtug** [comment]
@lminiero didn't get much further after fixing that. Now I'm getting deadlocked in this plugin creating a 2nd room. Looking like there are mutex leaks many places.

**Comment by tugtugtug** [comment]
looks like there are still lots to fix.
gateway->close_pc(p->session->handle) is called with rooms_mutex and audiobridge->mutex ,and the callback to janus_audiobridge_hangup_media will also then lock up the 2 mutexes.

**Comment by tugtugtug** [comment]
I'll park for now. Will come back to look at this further when I get a chance.

**Comment by lminiero** [comment]
Well, but this doesn't seem to have anything to do with the original issue, which was related to ignoring trickle candidates. Besides, your original steps mentioned the VideoRoom and not the AudioBridge. If that is fixed, then please let us know and we can consider that done. This other issue you highlighted can (and should) be tracked in a new issue, as I don't want us to start highjacking and confusing issues this way (it's already hard enough to track all issues as it is).

**Comment by tugtugtug** [comment]
@lminiero I'd love to verify the original problem, but there are just too many issues that block me to even reach that step with our test app at the moment. I just tried videoroom and it deadlocks right away after joining a newly created room.

janus_videoroom_participant_joining double locks the videoroom->mutex locked in the handler.

[plugins/janus_videoroom.c:janus_videoroom_handler:3510:140736674653952:lock] 0x7fffd8002470
[plugins/janus_videoroom.c:janus_videoroom_participant_joining:1257:140736674653952:lock] 0x7fffd8002470

Could team do some sanity test of this branch? It just seems it is too buggy at the moment to worth creating tickets for.

**Comment by lminiero** [comment]
We're not experiencing these issues at all, so not sure what there's to sanity check :wink: 
We'll try and replicate in the next few days to see if we can spot obvious issues that need fixing.

**Comment by tugtugtug** [comment]
@lminiero that's strange, let me try again with a clean build, I vaguely remember I may have merged some changes from my own branch, that may have merged some master stuff into the branch.

**Comment by lminiero** [comment]
Sorry, missed the edits (Github doesn't notify about those, only new posts). Still can't replicate.

**Comment by lminiero** [comment]
PS: made many changes in the last two weeks so line numbers don't match anymore. Please try with the latest code and let us know if that's still an issue (with updated refs in case).

**Comment by tugtugtug** [comment]
@lminiero Just tried again, looks like the audiobridge is still with issues. Creating a new audioroom (non-persistent) and join fails. I didn't try further, I think the deadlocks are still there after destroying this room.
I can also easily reproduce the videoroom deadlock, but seems just related to the notify_joining flag set to true. I see the function lock the room mutex when now in the refcount branch janus_videoroom_handler already locking the room mutex for a much larger scope.

**Comment by tugtugtug** [comment]
I copied some quick dirty changes I made previously to get things working at https://pastebin.com/cz5a86Gu

**Comment by tugtugtug** [comment]
I think with these changes, I don't see the original problem, but our client has also changed to wait for the hangup event. I'll verify again, once you fix the locking issues officially and remove our wait on the hangup event.

**Comment by lminiero** [comment]
See changes made for #1109: please let me know if they allow you to check if the original issue you opened the issue for was indeed fixed.

**Comment by lminiero** [comment]
Any update? Still an issue?

**Comment by tugtugtug** [comment]
@lminiero It still happens, however slightly different, I see that if a new publish happens before the ice thread is cleaned up, instead of ice failure, it seems to just fail to publish. But that may be expected? Anyhow, I think with a workaround to wait for the hangup, it works okay. Also, I see that we have finally the renegotiation implemented, which means our use case will move to that anyway. I'm okay closing this.

**Comment by lminiero** [comment]
Ack, thx for the feedback, closing then :+1:
